merkle.go

```diff
@@ -124,30 +133,89 @@ func (t *Tree) AddLeaf(value []byte) error {
 	return lastCachingError
 }
 
-// Root returns the root of the tree or an error if the number of leaves added is not a power of 2.
-func (t *Tree) Root() ([]byte, error) {
+func nextOrEmptyLayer(l *layer) *layer {
+	if l.next != nil {
+		return l.next
+	}
+	return &layer{height: l.height + 1}
+}
+
+// Root returns the root of the tree.
+// If the tree is unbalanced (num. of leaves is not a power of 2) it will perform padding on-the-fly.
+func (t *Tree) Root() []byte {
+	root, _ := t.RootAndProof()
+	return root
+}
+
+// Proof returns a partial tree proving the membership of leaves that were passed in leavesToProve when the tree was
+// initialized. For a single proved leaf this is a standard merkle proof (one sibling per layer of the tree from the
+// leaves to the root, excluding the proved leaf and root).
+// If the tree is unbalanced (num. of leaves is not a power of 2) it will perform padding on-the-fly.
+func (t *Tree) Proof() [][]byte {
+	_, proof := t.RootAndProof()
+	return proof
+}
+
+// RootAndProof returns the root of the tree and a partial tree proving the membership of leaves that were passed in
+// leavesToProve when the tree was initialized. For a single proved leaf this is a standard merkle proof (one sibling
+// per layer of the tree from the leaves to the root, excluding the proved leaf and root).
+// If the tree is unbalanced (num. of leaves is not a power of 2) it will perform padding on-the-fly.
+func (t *Tree) RootAndProof() ([]byte, [][]byte) {
+	ephemeralProof := t.proof
+	var ephemeralNode node
 	l := t.baseLayer
-	for {
-		if l.next == nil {
-			return l.parking.value, nil
+	for height := uint(0); height < t.minHeight || l != nil; height++ {
+
+		// If we've reached the last layer and the ephemeral node is still empty, the tree is balanced and the parked
+		// node is its root.
+		// In any other case (minHeight not reached, or the tree is unbalanced) we want to add padding at this point.
+		reachedMinHeight := height >= t.minHeight
+		onLastLayer := l != nil && l.next == nil
+		parkingIsBalancedTreeRoot := reachedMinHeight && onLastLayer && ephemeralNode.IsEmpty()
+		if parkingIsBalancedTreeRoot {
+			return l.parking.value, ephemeralProof
 		}
-		if l.parking.value != nil {
-			return nil, ErrorIncompleteTree
+
+		var parking node
+		if l != nil {
+			parking = l.parking
+		}
+		parent, lChild, rChild := t.calcEphemeralParent(parking, ephemeralNode)
+
+		// Consider adding children to the ephemeralProof.
+		if parent.onProvenPath {
```

Comment:
I don't see it... The default is very purposefully `false`. If for some reason it wasn't, I'd have created a method `onProvenPath()` that checks it internally. I'm happy that I don't need to do it.

Reply:
the thing is the `calcEphemeralParent` might return an empty node, in such case you should rule out the node because it is empty and not because it is not on the path to a proven leaf. Up to you.